File: packages/shared-types/src/legal.ts
/**
 * Legal and privacy types — consent records, cookie preferences,
 * age verification, and GDPR data export shapes.
 */

// ── Consent ──────────────────────────────────────────────────────────────

export type ConsentType = "terms" | "privacy" | "cookies" | "age_verification";

export interface ConsentRecord {
  id: string;
  consentType: ConsentType;
  version: string;
  acceptedAt: string; // ISO 8601
}

// ── Cookies ──────────────────────────────────────────────────────────────

export interface CookiePreferences {
  essential: true; // always true, read-only
  analytics: boolean;
}

// ── Age Verification ─────────────────────────────────────────────────────

/**
 * @future Pre-declared for future use when client-side parsing of age verification records is needed.
 */
export interface AgeVerification {
  confirmed: boolean;
  confirmedAt: string; // ISO 8601
}

// ── GDPR Data Export ─────────────────────────────────────────────────────

/**
 * @future Pre-declared for future use when client-side parsing of GDPR data export responses is needed.
 *
 * Represents the parsed GDPR data export structure in camelCase.
 *
 * Note: The API returns snake_case JSON and the current legal-client
 * returns a Blob (not parsed JSON). This type represents the parsed
 * structure for use if client-side parsing is ever added. The snake_case
 * API response fields (user_id, created_at, etc.) would need to be
 * transformed to match these camelCase fields.
 */
export interface DataExportResponse {
  userId: string;
  email: string;
  name: string;
  tier: string;
  createdAt: string;
  consents: ConsentRecord[];
  payments: Array<{
    id: string;
    amount: number;
    currency: string;
    status: string;
    createdAt: string;
  }>;
  cookiePreferences: {
    essential: true;
    analytics: boolean;
    updatedAt: string;
  } | null;
  sessions: Array<{
    id: string;
    ipAddress: string | null;
    userAgent: string | null;
    createdAt: string;
    expiresAt: string;
  }>;
  analyses: Array<{
    id: string;
    label: string;
    createdAt: string;
    updatedAt: string;
  }>;
  analysisCount: number;
  exportedAt: string;
}

File: apps/api/app/schemas/legal.py
"""
Legal and privacy schemas — request and response models for consent,
cookie preferences, and GDPR data export endpoints.
"""

from __future__ import annotations

from pydantic import BaseModel, Field

# ── Requests ──────────────────────────────────────────────────────────────


class RecordConsentRequest(BaseModel):
    """Record a consent event (terms, privacy, cookies, age_verification, genetic_data_processing)."""

    consent_type: str = Field(
        ...,
        pattern=r"^(terms|privacy|cookies|age_verification|genetic_data_processing)$",
        description="Type of consent being recorded",
    )
    version: str = Field(
        ...,
        min_length=1,
        max_length=20,
        description="Document version, e.g. '1.0'",
    )


class CookiePreferencesRequest(BaseModel):
    """Update cookie preferences."""

    analytics: bool = Field(
        default=False,
        description="Whether analytics cookies are enabled",
    )
    marketing: bool = Field(
        default=False,
        description="Whether marketing cookies are enabled",
    )


# ── Responses ─────────────────────────────────────────────────────────────


class ConsentResponse(BaseModel):
    """A single consent record."""

    id: str
    consent_type: str
    version: str
    accepted_at: str = Field(
        ...,
        description="ISO 8601 timestamp of when consent was given",
    )


class CookiePreferencesResponse(BaseModel):
    """Current cookie preferences for the user."""

    essential: bool = Field(
        default=True,
        description="Essential cookies are always enabled (read-only)",
    )
    analytics: bool
    marketing: bool = Field(
        default=False,
        description="Whether marketing cookies are enabled",
    )


# ── GDPR Export Sub-Schemas ──────────────────────────────────────────────


class ConsentExportSchema(BaseModel):
    """A consent record as it appears in the GDPR data export."""

    id: str
    consent_type: str
    version: str
    accepted_at: str
    ip_address: str | None = None
    user_agent: str | None = None


class PaymentExportSchema(BaseModel):
    """A payment record as it appears in the GDPR data export."""

    id: str
    amount: int
    currency: str
    status: str
    created_at: str


class SessionExportSchema(BaseModel):
    """An active session as it appears in the GDPR data export."""

    id: str
    ip_address: str | None = None
    user_agent: str | None = None
    created_at: str
    expires_at: str


class AnalysisExportSchema(BaseModel):
    """Analysis result metadata as it appears in the GDPR data export.

    Excludes encrypted result_data for security.
    """

    id: str
    label: str
    created_at: str
    updated_at: str


class CookiePreferencesExportSchema(BaseModel):
    """Cookie preferences as they appear in the GDPR data export."""

    essential: bool = True
    analytics: bool
    marketing: bool = False
    updated_at: str


class DataExportResponse(BaseModel):
    """GDPR data export — full dump of user data in JSON format.

    Note: Raw analysis data is NOT included because it is
    AES-256-GCM encrypted at rest. Only metadata and analysis
    count are returned.
    """

    user_id: str
    email: str
    name: str
    tier: str
    created_at: str
    consents: list[ConsentExportSchema]
    payments: list[PaymentExportSchema]
    cookie_preferences: CookiePreferencesExportSchema | None = Field(
        default=None,
        description="Current cookie preferences, or null if never set",
    )
    sessions: list[SessionExportSchema]
    analyses: list[AnalysisExportSchema]
    analysis_count: int
    exported_at: str
